### ISSUE TITLE
RBM score. There might be an inconsistency in parameter expectations

### ISSUE THREAD

**Issue by vlovsky** [deeplearning4j/deeplearning4j#3049]
RBM calculates scores by calling this https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/feedforward/rbm/RBM.java#L192 (note that the values are passed not the probabilities)

and then this https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/BasePretrainNetwork.java#L133

with respect to KL_DIVERGENCE loss function this is handled as https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossKLD.java#L21-L35

Also, recall that the KL divergence is:
![image](https://cloud.githubusercontent.com/assets/26049325/23951605/e52d59b0-0964-11e7-8dde-5101f76abb03.png)

Base on this I have two questions. 

1. Is it normal for the https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/feedforward/rbm/RBM.java#L192 to use negVSamples instead of negVProb? Do all loss functions expect probabilities in parameters?

2. Why would we need to execute the activation on Z in https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossKLD.java#L22 if Z is already an output of the RBM layer?

**Comment by vlovsky** [comment]
Theoretically KLD should be converging since we are doing Gibbs samples. It's not really happening with the standard KL_DIVERGENCE loss function. I ran few more tests with standard and custom loss functions. 

In the first example you can observe that the KLD didn't fully converge after 1000 iterations even though the input = output and input ~ probabilities. Also, you can see that the KLD fluctuates significantly (which theoretically should not be happening): https://gist.github.com/anonymous/914cf199af3d6e55ee1215bea3f0682f

In the next test I used custom loss function where I removed activation and used probabilities for z. It's very clear that the score is steadily converges (as it should for KLD) as the probabilities get closer to the empirical values.
https://gist.github.com/anonymous/f21ad3fd1bb630c76daac7a2a293a899

**Comment by eraly** [comment]
RBMs need to be scrubbed and it is quite likely that RBMs will need an overhaul...They have not been very high on the priority list so far. I am looking at them and will continue to do so. To answer your questions - 

Not all loss functions expect probabilities. MSE, MAE etc etc. I am unsure why the KLD loss function is even set on the RBMs since they are supposed to pretrain with constrastive divergence.

**Comment by vlovsky** [comment]
The best loss function I found so far is the squared loss, and since RBM has both input and output it just needs something that computes it without doing any conversion on the parameters.

**Comment by eraly** [comment]
We are dropping support for RBMs in favour of newer and more effective methods. Please take a look at variational autoencoders 
https://github.com/deeplearning4j/dl4j-examples/tree/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/unsupervised/variational
